synapse/push/pusherpool.py

```diff
@@ -346,17 +346,22 @@ async def _start_pusher(self, pusher_config: PusherConfig) -> Optional[Pusher]:
             )
             return None
 
-        if not p:
+        if not pusher:
             return None
 
         appid_pushkey = "%s:%s" % (pusher_config.app_id, pusher_config.pushkey)
 
         byuser = self.pushers.setdefault(pusher_config.user_name, {})
         if appid_pushkey in byuser:
-            byuser[appid_pushkey].on_stop()
-        byuser[appid_pushkey] = p
+            previous_pusher = byuser[appid_pushkey]
+            previous_pusher.on_stop()
 
-        synapse_pushers.labels(type(p).__name__, p.app_id).inc()
+            synapse_pushers.labels(
+                type(previous_pusher).__name__, previous_pusher.app_id
+            ).dec()
```

Comment:
They ought to cancel most of the time - the `app_ids` must be the same, otherwise `appid_pushkey` would be different.

However, I can't convince myself that they always cancel. I'm not sure that `type(previous_pusher).__name__` will always be the same as `type(pusher).__name__`.